services/src/main/java/org/keycloak/services/util/DefaultClientSessionContext.java

```diff
@@ -246,34 +230,26 @@ private boolean isClientScopePermittedForUser(ClientScopeModel clientScope) {
     private Set<RoleModel> loadRoles() {
         UserModel user = clientSession.getUserSession().getUser();
         ClientModel client = clientSession.getClient();
-
-        Set<ClientScopeModel> clientScopes = getClientScopes();
-
-        return TokenManager.getAccess(user, client, clientScopes);
+        return TokenManager.getAccess(user, client, getClientScopesStream());
     }
 
 
     private Set<ProtocolMapperModel> loadProtocolMappers() {
```

Comment:
This seems to be used only in line 140 above, and that method gets the Set returned here and calls .stream() on it to be able to return the proper stream type. Shouldn't this one simply return the stream?

Reply:
Same approach is used with roles and client scopes. My understanding is that we cache those entities to prevent multiple calls of `getClientScopesStream()` and `getProtocolMappersStream()` using this approach.

Reply:
Ah, it is indeed cached, this is fine then.